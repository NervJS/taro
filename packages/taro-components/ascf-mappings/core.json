{
  "Button": {
    "size": {
      "ascfProp": "dimension",
      "valueMap": {
        "mini": "XS",
        "small": "S",
        "default": "M",
        "large": "L"
      },
      "required": false,
      "default": "M",
      "docLink": "https://ascf.docs/components/button#dimension"
    },
    "loading": {
      "ascfProp": "isLoading",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/button#isloading"
    }
  },
  "Input": {
    "password": {
      "ascfProp": "secretInput",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/input#secretinput"
    }
  }
}
