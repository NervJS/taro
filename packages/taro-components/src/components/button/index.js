import 'weui'
import Nerv from 'nervjs'
import omit from 'omit.js'
import classNames from 'classnames'

import './style/index.scss'

class Button extends Nerv.Component {
  constructor () {
    super(...arguments)
    this.state = {
      hover: false,
      touch: false
    }
  }

  componentWillUnmount () {
    this.startTimer && clearTimeout(this.startTimer)
    this.endTimer && clearTimeout(this.endTimer)
  }

  render () {
    const {
      children,
      disabled,
      className,
      style,
      onClick,
      onTouchStart,
      onTouchEnd,
      hoverClass = 'button-hover',
      hoverStartTime = 20,
      hoverStayTime = 70,
      size,
      plain,
      loading = false,
      type = 'default'
    } = this.props
    const cls = className || classNames(
      'weui-btn',
      {
        [`${hoverClass}`]: this.state.hover && !disabled && hoverClass !== 'none',
        [`weui-btn_plain-${type}`]: plain,
        [`weui-btn_${type}`]: !plain && type,
        'weui-btn_mini': size === 'mini',
        'weui-btn_loading': loading,
        'weui-btn_disabled': disabled
      }
    )

    const _onTouchStart = e => {
      this.setState(() => ({
        touch: true
      }))
      if (hoverClass && hoverClass !== 'none' && !disabled) {
        this.startTimer = setTimeout(() => {
          if (this.state.touch) {
            this.setState(() => ({
              hover: true
            }))
          }
        }, hoverStartTime)
      }
      onTouchStart && onTouchStart(e)
    }
    const _onTouchEnd = e => {
      this.setState(() => ({
        touch: false
      }))
      if (hoverClass && hoverClass !== 'none' && !disabled) {
        this.endTimer = setTimeout(() => {
          if (!this.state.touch) {
            this.setState(() => ({
              hover: false
            }))
          }
        }, hoverStayTime)
      }
      onTouchEnd && onTouchEnd(e)
    }

    return (
      <button
        {...omit(this.props, ['hoverClass', 'onTouchStart', 'onTouchEnd'])}
        className={cls}
        style={style}
        onClick={onClick}
        disabled={disabled}
        onTouchStart={_onTouchStart}
        onTouchEnd={_onTouchEnd}
      >
        {loading && <i class='weui-loading' />}
        {children}
      </button>
    )
  }
}

export default Button
