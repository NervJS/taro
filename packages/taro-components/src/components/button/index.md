button

## API

|     | 属性                   | 类型    | 默认值       | 说明                                                                                        |
| --- | ---------------------- | ------- | ------------ | ------------------------------------------------------------------------------------------- |
| √   | type                   | String  | default      | 按钮的样式类型                                                                              |
| √   | size                   | String  | default      | 按钮的大小 px                                                                               |
| √   | plain                  | Boolean | false        | 按钮是否镂空，背景色透明                                                                    |
| √   | disabled               | Boolean | false        | 是否禁用                                                                                    |
| √   | loading                | Boolean | false        | 名称前是否带 loading 图标                                                                   |
|     | form-type              | String  |              | 用于 form 组件，点击分别会触发 form 组件的 submit/reset 事件                                |
|     | open-type              | String  |              | 微信开放能力                                                                                |
|     | app-parameter          | String  |              | 打开 APP 时，向 APP 传递的参数                                                              |
| √   | hover-class            | String  | button-hover | 指定按钮按下去的样式类。当 hover-class="none" 时，没有点击态效果                            |
|     | hover-stop-propagation | Boolean | false        | 指定是否阻止本节点的祖先节点出现点击态                                                      |
| √   | hover-start-time       | Number  | 20           | 按住后多久出现点击态，单位毫秒                                                              |
| √   | hover-stay-time        | Number  | 70           | 手指松开后点击态保留时间，单位毫秒                                                          |
|     | bindgetuserinfo        | Handler |              | 用户点击该按钮时，会返回获取到的用户信息，从返回参数的 detail 中获取到的值同 wx.getUserInfo |
|     | lang                   | String  | en           | 指定返回用户信息的语言，zh_CN 简体中文，zh_TW 繁体中文，en 英文。                           |
