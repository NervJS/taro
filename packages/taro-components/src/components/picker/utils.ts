function getTimeRange (begin: number, end: number) {
  const range: string[] = []
  for (let i = begin; i <= end; i++) {
    range.push(`${i < 10 ? '0' : ''}${i}`)
  }
  return range
}

export const hoursRange = [
  '20',
  '21',
  '22',
  '23',
  ...getTimeRange(0, 23),
  '00',
  '01',
  '02',
  '03'
]

export const minutesRange = [
  '56',
  '57',
  '58',
  '59',
  ...getTimeRange(0, 59),
  '00',
  '01',
  '02',
  '03'
]

/**
 * 校验传入的 value 是否合法
 */
export function verifyValue (value: number, range: any[]) {
  if (!isNaN(+value) && value >= 0 && value < range.length) return true
  return false
}

/**
 * 检验传入的 time value 是否合法
 */
export function verifyTime (value: string) {
  if (!/^\d{1,2}:\d{1,2}$/.test(value)) return false

  const time = value.split(':').map(num => +num)

  if (time[0] < 0 || time[0] > 23) return false
  if (time[1] < 0 || time[1] > 59) return false

  return true
}

/**
 * 比较时间
 * return t1 <= t2
 */
export function compareTime (t1: string, t2: string) {
  const t1List = t1.split(':').map(i => +i)
  const t2List = t2.split(':').map(i => +i)

  if (t1List[0] < t2List[0]) return true
  if (t1List[0] === t2List[0] && t1List[1] <= t2List[1]) return true

  return false
}

/**
 * 校验日期合法性，返回合法性和日期数组
 */
export function verifyDate (dateStr: string) {
  if (!dateStr) return false
  const date = new Date(dateStr.replace(/-/g, '/'))
  return isNaN(date.getMonth()) ? false : date
}

/**
 * 获取当月最大天数
 */
function getMaxDay (year: number, month: number) {
  if (month === 4 || month === 6 || month === 9 || month === 11) return 30
  if (month === 2) {
    if ((year % 4 === 0 && year % 100 !== 0) || year % 400 === 0) return 29
    else return 28
  }
  return 31
}

export function formatValue (value: number | number[] | string) {
  let res: number | string | string[]
  if (Array.isArray(value)) {
    res = value.map(item => String(item))
  } else {
    res = value
  }
  return res
}

/**
 * 获取时间数组
 */
function getDateRange (start: number, end: number) {
  const range: number[] = []
  for (let i = start; i <= end; i++) {
    range.push(i)
  }
  return range
}

/**
 * 获取年份区间数组
 */
export function getYearRange (start: number, end: number) {
  return getDateRange(start, end)
}

/**
 * 获取月份区间数组
 */
export function getMonthRange (start: Date, end: Date, year: number) {
  let rangeStart = 1
  let rangeEnd = 12

  // 当前年份等于开始年份，由开始对应的月份约束开始值
  if (start.getFullYear() === year) {
    rangeStart = start.getMonth() + 1
  }

  // 当前年份等于结束年份，由结束对应的月份约束结束值
  if (end.getFullYear() === year) {
    rangeEnd = end.getMonth() + 1
  }

  return getDateRange(rangeStart, rangeEnd)
}

/**
 * 获取日期区间数组
 */
export function getDayRange (start: Date, end: Date, year: number, month: number) {
  let rangeStart = 1
  let rangeEnd = getMaxDay(year, month)

  if (start.getFullYear() === year && start.getMonth() + 1 === month) {
    rangeStart = start.getDate()
  }

  if (end.getFullYear() === year && end.getMonth() + 1 === month) {
    rangeEnd = end.getDate()
  }

  return getDateRange(rangeStart, rangeEnd)
}
