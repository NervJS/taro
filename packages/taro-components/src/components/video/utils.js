export const formatTime = time => {
  if (time === null) return ''
  const sec = Math.round(time % 60)
  const min = Math.round((time - sec) / 60)
  return `${min < 10 ? `0${min}` : min}:${sec < 10 ? `0${sec}` : sec}`
}

export const calcDist = (x, y) => {
  return Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2))
}

export const normalizeNumber = number => {
  return Math.max(-1, Math.min(number, 1))
}
