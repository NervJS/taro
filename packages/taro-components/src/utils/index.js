export function throttle (fn, threshhold, scope) {
  threshhold || (threshhold = 250)
  let last, deferTimer
  return function () {
    let context = scope || this

    let now = +new Date()
    let args = arguments
    if (last && now < last + threshhold) {
      clearTimeout(deferTimer)
      deferTimer = setTimeout(() => {
        last = now
        fn.apply(context, args)
      }, threshhold)
    } else {
      last = now
      fn.apply(context, args)
    }
  }
}

export const normalizePath = url => {
  let _isRelative
  let _leadingParents = ''
  let _parent, _pos

  // handle relative paths
  if (url.charAt(0) !== '/') {
    _isRelative = true
    url = '/' + url
  }

  // handle relative files (as opposed to directories)
  if (url.substring(-3) === '/..' || url.slice(-2) === '/.') {
    url += '/'
  }

  // resolve simples
  url = url.replace(/(\/(\.\/)+)|(\/\.$)/g, '/').replace(/\/{2,}/g, '/')

  // remember leading parents
  if (_isRelative) {
    _leadingParents = url.substring(1).match(/^(\.\.\/)+/) || ''
    if (_leadingParents) {
      _leadingParents = _leadingParents[0]
    }
  }

  // resolve parents
  while (true) {
    _parent = url.search(/\/\.\.(\/|$)/)
    if (_parent === -1) {
      // no more ../ to resolve
      break
    } else if (_parent === 0) {
      // top level cannot be relative, skip it
      url = url.substring(3)
      continue
    }

    _pos = url.substring(0, _parent).lastIndexOf('/')
    if (_pos === -1) {
      _pos = _parent
    }
    url = url.substring(0, _pos) + url.substring(_parent + 3)
  }

  // revert to relative
  if (_isRelative) {
    url = _leadingParents + url.substring(1)
  }

  return url
}

export const splitUrl = _url => {
  let url = _url || ''
  let pos
  let res = {
    path: null,
    query: null,
    fragment: null
  }

  pos = url.indexOf('#')
  if (pos > -1) {
    res.fragment = url.substring(pos + 1)
    url = url.substring(0, pos)
  }

  pos = url.indexOf('?')
  if (pos > -1) {
    res.query = url.substring(pos + 1)
    url = url.substring(0, pos)
  }

  res.path = url

  return res
}

export const isNumber = obj => {
  return typeof obj === 'number'
}
