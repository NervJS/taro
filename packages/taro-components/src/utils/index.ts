export function throttle (fn, ms = 250, scope?) {
  let last, deferTimer
  return function () {
    const context = scope || this

    const now = +new Date()
    // eslint-disable-next-line prefer-rest-params
    const args = arguments
    if (last && now < last + ms) {
      clearTimeout(deferTimer)
      deferTimer = setTimeout(() => {
        last = now
        fn.apply(context, args)
      }, ms)
    } else {
      last = now
      fn.apply(context, args)
    }
  }
}

export function debounce (fn, ms = 250) {
  let timer: NodeJS.Timeout

  return function (...arrs) {
    clearTimeout(timer)
    timer = setTimeout(function () {
      fn(...arrs)
    }, ms)
  }
}

export * from './style'
export * from './url'
