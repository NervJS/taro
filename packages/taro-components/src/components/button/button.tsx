// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { Component, h, Prop, State, ComponentInterface, Event, EventEmitter, Listen, Element, Host } from '@stencil/core'
import classNames from 'classnames'

@Component({
  tag: 'taro-button-core',
  styleUrl: './style/index.scss'
})
export class Button implements ComponentInterface {
  @Prop() disabled: boolean
  @Prop() hoverClass = 'button-hover'
  @Prop() type = ''
  @Prop() hoverStartTime = 20
  @Prop() hoverStayTime = 70
  @Prop() size: string
  @Prop() plain: boolean
  @Prop() loading = false
  @Prop() formType: 'submit' | 'reset' | null = null

  @Element() el: HTMLElement

  @State() hover = false
  @State() touch = false

  @Event({
    eventName: 'tarobuttonsubmit'
  }) onSubmit: EventEmitter

  @Event({
    eventName: 'tarobuttonreset'
  }) onReset: EventEmitter

  @Listen('touchstart')
  onTouchStart () {
    if (this.disabled) {
      return
    }

    this.touch = true
    if (this.hoverClass && !this.disabled) {
      setTimeout(() => {
        if (this.touch) {
          this.hover = true
        }
      }, this.hoverStartTime)
    }
  }

  @Listen('touchend')
  onTouchEnd () {
    if (this.disabled) {
      return
    }

    this.touch = false
    if (this.hoverClass && !this.disabled) {
      setTimeout(() => {
        if (!this.touch) {
          this.hover = false
        }
      }, this.hoverStayTime)
    }

    if (this.formType === 'submit') {
      this.onSubmit.emit()
    } else if (this.formType === 'reset') {
      this.onReset.emit()
    }
  }

  render () {
    const {
      disabled,
      hoverClass,
      type,
      size,
      plain,
      loading,
      hover
    } = this

    const cls = classNames({
      [`${hoverClass}`]: hover && !disabled
    })

    return (
      <Host
        class={cls}
        type={type}
        plain={plain}
        loading={loading}
        size={size}
        disabled={disabled}
      >
        {loading && <i class='weui-loading' />}
        <slot />
      </Host>
    )
  }
}
