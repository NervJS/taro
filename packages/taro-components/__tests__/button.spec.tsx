import { h } from '@stencil/core'
import { AnyHTMLElement } from '@stencil/core/internal'
import { newSpecPage, SpecPage } from '@stencil/core/testing'

import { Button } from '../src/components/button/button'
import { delay } from './utils'

describe('Button', () => {
  let page: SpecPage
  it('props', async () => {
    const size = 'mini'
    const plain = true
    const loading = true
    const disabled = false
    page = await newSpecPage({
      components: [Button],
      template: () => (<taro-button-core
        size={size}
        plain={plain}
        loading={loading}
        disabled={disabled}
      >button</taro-button-core>),
    })

    expect(page.root?.type).not.toBeUndefined()
    expect(page.root?.plain).toEqual(true)
    expect(page.root?.loading).toEqual(true)
    expect(page.root?.size).toEqual('mini')
    expect(page.root?.disabled).toEqual(false)

    await delay(3000)
    const icon = page.root?.querySelector<AnyHTMLElement>('i')
    expect(icon?.classList.contains('weui-loading')).toEqual(true)
    expect(page.root?.textContent).toEqual('button')

    page.root?.setAttribute('plain', 'false')
    await page.waitForChanges()
    expect(page.root?.plain).toEqual(false)

    page.root?.setAttribute('loading', 'false')
    await page.waitForChanges()
    expect(page.root?.loading).toEqual(false)
    expect(icon?.parentNode).toBeNull()

    page.root?.setAttribute('disabled', 'true')
    await page.waitForChanges()
    expect(page.root?.disabled).toEqual(true)

    page.root?.setAttribute('size', 'big')
    await page.waitForChanges()
    expect(page.root?.size).toEqual('big')

    expect(page.root).toMatchSnapshot()
  })
})
