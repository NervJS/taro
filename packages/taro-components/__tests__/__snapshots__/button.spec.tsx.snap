// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button props 1`] = `
<taro-button-core disabled="" loading="false" plain="false" size="big" type="">
  <!---->
  button
</taro-button-core>
`;
